pkg/filestore/writer_test.go

```diff
@@ -299,6 +315,91 @@ func Test_WriteKeypair(t *testing.T) {
+			testBundle: pkcs8Bundle,
+			meta: metadata.Metadata{
+				VolumeID:   "vol-id",
+				TargetPath: "/target-path",
+				VolumeContext: map[string]string{
+					"csi.cert-manager.io/issuer-name":              "ca-issuer",
+					"csi.cert-manager.io/key-encoding":             "PKCS8",
+					"csi.cert-manager.io/pkcs12-enable":   "true",
+					"csi.cert-manager.io/pkcs12-filename":     "my-file.pfx",
+					"csi.cert-manager.io/pkcs12-password": "my-password",
+				},
+			},
+			expFiles: map[string][]byte{
+				"ca.crt":  pkcs8Bundle.caPEM,
+				"tls.crt": pkcs8Bundle.certPEM,
+				"tls.key": pkcs8Bundle.pkPEM,
+				"metadata.json": []byte(
+					`{"volumeID":"vol-id","targetPath":"/target-path","nextIssuanceTime":"1970-01-03T00:00:00Z","volumeContext":{"csi.cert-manager.io/issuer-name":"ca-issuer","csi.cert-manager.io/key-encoding":"PKCS8","csi.cert-manager.io/pkcs12-enable":"true","csi.cert-manager.io/pkcs12-filename":"my-file.pfx","csi.cert-manager.io/pkcs12-password":"my-password"}}`,
+				),
+			},
+			expErr: false,
+		},
+		"keystore PKCS12 with no file should default to ketstore.p12": {
+			testBundle: pkcs8Bundle,
+			meta: metadata.Metadata{
+				VolumeID:   "vol-id",
+				TargetPath: "/target-path",
+				VolumeContext: map[string]string{
+					"csi.cert-manager.io/issuer-name":              "ca-issuer",
+					"csi.cert-manager.io/key-encoding":             "PKCS8",
+					"csi.cert-manager.io/pkcs12-enable":   "true",
+					"csi.cert-manager.io/pkcs12-password": "my-password",
+				},
+			},
+			expFiles: map[string][]byte{
+				"ca.crt":  pkcs8Bundle.caPEM,
+				"tls.crt": pkcs8Bundle.certPEM,
+				"tls.key": pkcs8Bundle.pkPEM,
+				"metadata.json": []byte(
+					`{"volumeID":"vol-id","targetPath":"/target-path","nextIssuanceTime":"1970-01-03T00:00:00Z","volumeContext":{"csi.cert-manager.io/issuer-name":"ca-issuer","csi.cert-manager.io/key-encoding":"PKCS8","csi.cert-manager.io/pkcs12-enable":"true","csi.cert-manager.io/pkcs12-password":"my-password"}}`,
+				),
+			},
+			expErr: false,
+		},
+		"keystore PKCS12 with no password and breakout file should error": {
+			testBundle: pkcs8Bundle,
+			meta: metadata.Metadata{
+				VolumeID:   "vol-id",
+				TargetPath: "/target-path",
+				VolumeContext: map[string]string{
+					"csi.cert-manager.io/issuer-name":              "ca-issuer",
+					"csi.cert-manager.io/key-encoding":             "PKCS8",
+					"csi.cert-manager.io/pkcs12-enable":   "true",
+					"csi.cert-manager.io/pkcs12-filename":     "../my-file.pfx",
+					"csi.cert-manager.io/pkcs12-password": "",
+				},
+			},
+			expFiles: map[string][]byte{
+				"metadata.json": []byte(
+					`{"volumeID":"vol-id","targetPath":"/target-path","volumeContext":{"csi.cert-manager.io/issuer-name":"ca-issuer","csi.cert-manager.io/key-encoding":"PKCS8","csi.cert-manager.io/pkcs12-enable":"true","csi.cert-manager.io/pkcs12-filename":"../my-file.pfx","csi.cert-manager.io/pkcs12-password":""}}`,
+				),
+			},
+			expErr: true,
```

Comment:
**question (blocking)**: if an error occurs, no files should be written - correct? 
we should not expect `metadata.json` here

**suggestion**: use this change from my PR: https://github.com/cert-manager/csi-driver/pull/101/files#diff-500bda06db839eb31b664ea0abc0e4b7808ff7c4fd91ed8caa77f9f6f2273d82R391

Reply:
We call `RegisterMetadata` at the start of every test, which is one of the first things we do when we receive a `NodePublishVolume` https://github.com/cert-manager/csi-lib/blob/8e344a2d37de27bacb17f8e1d5ddc1f03733351a/driver/nodeserver.go#L72

Like elsewhere, the `nextIssuanceTime` being a non-nil value can be checked to see whether there was a successful issuance. Even if we expect an error, we want to check the state of the metadata file.